back/backend/load_db/load_db.py

```diff
@@ -0,0 +1,195 @@
+                    if 'id' in row:
+                        fields_to_write += ['id']
+                    to_write = {}
+                    for field in fields_to_write:
+                        if field in row:
+                            to_write[field] = row[field]
+                    row_id = cls.get_id(row)
+                    if hash(row_id) % shards == shard:
+                        out[row_id] = to_write
+    except FileNotFoundError as e:
+        # This is fine, the file might just not exist.
+        logging.exception(e)
+        pass
+    return out
+
+
+def write_table(path, rows, cls, shard):
+    use_id = False
+    with open(path, 'r') as csvfile:
+        reader = csv.DictReader(csvfile, dialect='excel-tab')
+        use_id = 'id' in reader.fieldnames
+    with open(path + '.filtered', 'w' if shard == 0 else 'a') as csvfile:
+        fields_to_write = cls.columns_used()
+        if use_id:
+            fields_to_write += ['id']
+        writer = csv.DictWriter(csvfile, dialect='excel-tab', fieldnames=fields_to_write)
+        if shard == 0:
+            writer.writeheader()
+        for row in rows.items():
+            writer.writerow({k: v for k, v in row[1].items() if k in fields_to_write})
+
+
+def process_export(old_export_path, new_export_path):
+    client = ndb.Client()
+    for table, cls, shards in get_tables():
+        logging.info('Processing ' + table)
+        table_suffix = '/WCA_export_' + table + '.tsv'
+        for shard in range(shards):
+            logging.info('Shard %d/%d' % (shard + 1, shards))
+            old_rows = read_table(old_export_path + table_suffix + '.filtered', cls, False, shard, shards)
+            new_rows = read_table(new_export_path + table_suffix, cls, True, shard, shards)
+            logging.info('Old: %d' % len(old_rows))
+            logging.info('New: %d' % len(new_rows))
+            write_table(new_export_path + table_suffix, new_rows, cls, shard)
+
+            objects_to_put = []
+            keys_to_delete = []
+
+            modifier = None
+            with client.context():
+                modifier = get_modifier(table)
+            for key in new_rows:
+                row = new_rows[key]
+                if key in old_rows and old_rows[key] == row:
+                    continue
+                else:
+                    with client.context():
+                        obj = cls(id=key)
+                        obj.parse_from_dict(row)
+                        if modifier:
+                            modifier(obj)
+                        objects_to_put += [obj]
+            for key, row in old_rows.items():
+                if key in new_rows:
+                    continue
+                else:
+                    with client.context():
+                        keys_to_delete += [ndb.Key(cls, key)]
+
+            logging.info('Putting %d objects' % len(objects_to_put))
+            while objects_to_put:
+                with client.context():
```

Comment:
Is context meant to in the while statement? Seems like a lot of extra initialization with a tight loop